arrows/core/generate_texture.h

```diff
@@ -0,0 +1,427 @@
+ */
+template <class T>
+void render_triangle_from_images(vital::vector_2d const& v1, vital::vector_2d const& v2, vital::vector_2d const& v3,
+                                 vital::vector_3d const& pt1, vital::vector_3d const& pt2, vital::vector_3d const& pt3,
+                                 std::vector<vital::camera_sptr> const& cameras, std::vector<vital::image> const& images,
+                                 std::vector<double> const& depths_pt1,
+                                 std::vector<double> const& depths_pt2,
+                                 std::vector<double> const& depths_pt3,
+                                 std::vector<vital::image> const& depth_maps,
+                                 vital::image& texture, double depth_threshold,
+                                 image_fusion_method<T> const& adjust_images_contributions)
+{
+  // Compute a score for each image
+  std::vector<double> scores(images.size(), 0.0);
+  vital::matrix_3x3d points;
+  for (size_t i = 0; i < images.size(); ++i)
+  {
+    points << cameras[i]->project(pt1), cameras[i]->project(pt2), cameras[i]->project(pt3), 1, 1, 1;
+    scores[i] = std::max(-points.determinant(), 0.0);
+  }
+
+  triangle_scan_iterator tsi(v1, v2, v3);
+  vital::matrix_3x3d triangle;
+  vital::matrix_3x3d area_c, area_a;
+  triangle << v1, v2, v3, 1, 1, 1;
+  area_c << v1, v2, v2, 1, 1, 1;  // triangle abp
+  area_a << v2, v3, v3, 1, 1, 1;  // triangle bcp
+  double inv_area = 1.0 / triangle.determinant();
+  std::vector<vital::vector_2d> points_2d(images.size());
+  std::vector<T> values(texture.depth(), 0.0);
+  for (tsi.reset(); tsi.next(); )
+  {
+    int y = tsi.scan_y();
+    if (y < 0 || y >= static_cast<int>(texture.height()))
+      continue;
+    int min_x = std::max(0, tsi.start_x());
+    int max_x = std::min(static_cast<int>(texture.width()) - 1, tsi.end_x());
+
+    area_c(1, 2) = y;
+    area_a(1, 2) = y;
+    for (int x = min_x; x <= max_x; ++x)
+    {
+      area_c(0, 2) = x;
+      area_a(0, 2) = x;
+      vital::vector_3d bary_coords(area_a.determinant() * inv_area, 0, area_c.determinant() * inv_area);
+      bary_coords(1) = 1 - bary_coords(0) - bary_coords(2);
+      // Corresponding 3d point
+      vital::vector_3d pt3d = bary_coords(0) * pt1 + bary_coords(1) * pt2 + bary_coords(2) * pt3;
+
+      std::vector<double> point_scores = scores;
+      for (size_t i = 0; i < images.size(); ++i)
+      {
+        // Corresponding point in image i
+        vital::vector_2d pt_img = cameras[i]->project(pt3d);
+        points_2d[i] = pt_img;
+        // border check from the camera i
+        if (pt_img(0) < 0 || pt_img(0) >= images[i].width() || pt_img(1) < 0 || pt_img(1) >= images[i].height())
+        {
+          point_scores[i] = 0;
+          continue;
+        }
+        // visibility test from the camera i
+        double interpolated_depth = bary_coords(0) * depths_pt1[i] + bary_coords(1) * depths_pt2[i] + bary_coords(2) * depths_pt3[i];
```

Comment:
You probably don't need to pre-compute all the vertex depths for all the images and then interpolate them here.  Instead you could just project the interpolated, `pt3d`, with the camera depth function.